[test_attention_1d.py] | (T3K) T3000 e2e tests | t3000-e2e-tests / models_tttv2_attention_full_set_unit_tests [wh_llmbox]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] TT_FATAL in partition.cpp:152 causes 56 test_attention_1d_vs_reference failures on T3000 wh_llmbox
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** (T3K) T3000 e2e tests
**Job:** t3000-e2e-tests / models_tttv2_attention_full_set_unit_tests [wh_llmbox]

### Regression timeline

- **Last passing commit:** N/A (not provided)
- **First failing commit:** `09a0526ebfe5` — 2026-05-29T07:35:35Z — [run](https://github.com/tenstorrent/tt-metal/actions/runs/26624591673)
- **Most recent failing commit:** `039767d00b9a` — 2026-06-12T07:52:20Z — [run](https://github.com/tenstorrent/tt-metal/actions/runs/27402500504)

The regression has persisted for at least ~2 weeks.

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/27813747862/job/82319197711
- https://github.com/tenstorrent/tt-metal/actions/runs/27745215829/job/82083335749

### Failing test path(s)

- `models/common/tests/modules/attention/test_attention_1d.py` — `test_attention_1d_vs_reference` (56 parameterized variants)

**Test command:**
```
HF_HUB_OFFLINE=1 HF_HOME=/mnt/MLPerf/huggingface TT_CACHE_PATH=/mnt/MLPerf/huggingface/tt_cache/tttv2/attention_1d pytest --tb=short models/common/tests/modules/attention
```

### Error excerpt (terminal failure from logs)

**Run 1** (2026-06-12):
```
FAILED models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-2048-Qwen2-7B-1x2] - RuntimeError: TT_FATAL @ /project/ttnn/core/tensor/xtensor/partition.cpp:152: std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()
FAILED models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-4096-Qwen2-7B-1x2] - RuntimeError: TT_FATAL @ /project/ttnn/core/tensor/xtensor/partition.cpp:152: std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()
...
= 56 failed, 320 passed, 128 skipped, 948 deselected, 1 warning in 3396.87s (0:56:36) =
```

**Run 2** (2026-06-11):
```
FAILED models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-2048-Qwen2-7B-1x2] - RuntimeError: TT_FATAL @ /project/ttnn/core/tensor/xtensor/partition.cpp:152: std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()
FAILED models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-4096-Qwen2-7B-1x2] - RuntimeError: TT_FATAL @ /project/ttnn/core/tensor/xtensor/partition.cpp:152: std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()
...
= 56 failed, 320 passed, 128 skipped, 948 deselected, 1 warning in 3420.20s (0:57:00) =
```

The fatal assertion is: duplicate dimensions passed to the tensor partition function at `ttnn/core/tensor/xtensor/partition.cpp:152`.

### Reproduction steps

On a T3000 wormhole_b0 (wh_llmbox) host with the tt-metal environment set up:

```bash
HF_HUB_OFFLINE=1 \
HF_HOME=/mnt/MLPerf/huggingface \
TT_CACHE_PATH=/mnt/MLPerf/huggingface/tt_cache/tttv2/attention_1d \
pytest --tb=short models/common/tests/modules/attention
```

To target a specific failing case:
```bash
pytest --tb=short "models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-2048-Qwen2-7B-1x2]"
```

### Notes

- The assertion failure occurs in `ttnn/core/tensor/xtensor/partition.cpp:152` — the check `std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()` means duplicate dimension indices are being passed to the partition function.
- 56 test variants fail across multiple model configs (Qwen2-7B, Qwen2.5-7B, DeepSeek-R1-14B, Qwen2.5-Coder-32B) and KV-cache strategies (paged, paged-chunked, standard), all on the `1x2` and `1x8` mesh configs.
- The failure affects multi-device (1x2, 1x8) configurations but not single-device runs (those appear to pass).
- **Determinism check:**
  - Run 1 terminal error: `FAILED models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-2048-Qwen2-7B-1x2] - RuntimeError: TT_FATAL @ /project/ttnn/core/tensor/xtensor/partition.cpp:152: std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()`
  - Run 2 terminal error: `FAILED models/common/tests/modules/attention/test_attention_1d.py::test_attention_1d_vs_reference[10-paged-1x2-prefill-2048-Qwen2-7B-1x2] - RuntimeError: TT_FATAL @ /project/ttnn/core/tensor/xtensor/partition.cpp:152: std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end()`
  - Both runs: 56 failed, 320 passed, 128 skipped, 948 deselected — identical test counts and identical error class.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: (T3K) T3000 e2e tests`
`Auto-triage-job-name: t3000-e2e-tests / models_tttv2_attention_full_set_unit_tests [wh_llmbox]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: not found
Slack notification: not found
<!-- PIPELINE-REPORT-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

--- a/models/common/tests/modules/attention/test_attention_1d.py
+++ b/models/common/tests/modules/attention/test_attention_1d.py
@@ -43,7 +43,7 @@ from models.common.modules.rmsnorm.rmsnorm_1d import RMSNorm1DConfig
 from models.common.modules.tt_ccl import TT_CCL
 from models.common.tensor_utils import get_rot_transformation_mat, zeros_like_kv_cache, zeros_like_paged_cache
 from models.common.tests.utils import stable_model_seed
-from models.common.utility_functions import comp_allclose, comp_pcc, is_wormhole_b0, nearest_32
+from models.common.utility_functions import comp_allclose, comp_pcc, nearest_32
 
 # =============================================================================
 # RoPE Helper Functions (replaces TTTv1 rope imports)
@@ -1186,11 +1186,6 @@ def test_attention_1d_vs_reference(
     test_attention_1d_vs_reference_from_model_args which uses existing infrastructure
     that handles HF API differences.
     """
-    # WH B0 multi-device: reduce_scatter output topology generates duplicate shard dims
-    # causing TT_FATAL @ partition.cpp:152. Single-device (1x1) is unaffected. Refs #46878.
-    if is_wormhole_b0() and ttnn_mesh_device.get_num_devices() > 1:
-        pytest.skip("WH B0 multi-device: TT_FATAL duplicate dims in partition.cpp (refs #46878)")
-
     # Skip if mesh_shape doesn't match device
     device_shape = (ttnn_mesh_device.shape[0], ttnn_mesh_device.shape[1])
     if device_shape != mesh_shape:

--- a/ttnn/cpp/ttnn/operations/experimental/ccl/reduce_scatter_minimal_async/device/reduce_scatter_minimal_async_op_device_operation.cpp
+++ b/ttnn/cpp/ttnn/operations/experimental/ccl/reduce_scatter_minimal_async/device/reduce_scatter_minimal_async_op_device_operation.cpp
@@ -124,23 +124,44 @@ std::vector<Tensor> ReduceScatterMinimalAsyncDeviceOperation::create_output_tens
 
 std::vector<tt::tt_metal::TensorTopology> ReduceScatterMinimalAsyncDeviceOperation::compute_output_topologies(
     const ReduceScatterMinimalAsyncParams& operation_attributes, const ReduceScatterMinimalAsyncInputs& tensor_args) {
-    // reduce_scatter produces (intermediate, output). The output is sharded along `dim`
-    // across `cluster_axis`; the intermediate is an internal workspace whose topology is
-    // best left matching the input so that downstream introspection is not misleading.
+    // TODO(#48421): Enforce input invariants with TT_FATAL instead of sanitising malformed topologies.
+    // Output is sharded along `dim` across `cluster_axis`; intermediate keeps the input topology.
     const auto& input_topology = tensor_args.input_tensor.tensor_topology();
     auto output_placements = input_topology.placements();
 
-    auto shard_placement =
-        tt::tt_metal::distributed::MeshMapperConfig::Shard{static_cast<int>(operation_attributes.dim)};
+    // Normalize `dim` so Shard placements are stored consistently (avoids [Shard(-1), Shard(3)]).
+    const int rank = static_cast<int>(tensor_args.input_tensor.logical_shape().rank());
+    const int dim = ((static_cast<int>(operation_attributes.dim) % rank) + rank) % rank;
+    const auto shard_placement = tt::tt_metal::distributed::MeshMapperConfig::Shard{dim};
+
+    // Replicate any other mesh axis still sharding the same tensor dim — concat_ndim requires unique dims.
+    auto clear_same_dim = [&](auto& placement) {
+        if (auto* shard = std::get_if<tt::tt_metal::distributed::MeshMapperConfig::Shard>(&placement)) {
+            if (((shard->dim % rank) + rank) % rank == dim) {
+                placement = tt::tt_metal::distributed::MeshMapperConfig::Replicate{};
+            }
+        }
+    };
 
     if (operation_attributes.cluster_axis.has_value()) {
         const auto axis = operation_attributes.cluster_axis.value();
         if (axis < output_placements.size()) {
+            for (size_t i = 0; i < output_placements.size(); ++i) {
+                if (i != axis) {
+                    clear_same_dim(output_placements[i]);
+                }
+            }
             output_placements[axis] = shard_placement;
         }
     } else {
-        for (auto& placement : output_placements) {
-            placement = shard_placement;
+        // Whole-mesh reduce_scatter: shard only real (>1) axes on `dim`; replicate size-1 axes to keep dims unique.
+        const auto& mesh_shape = input_topology.distribution_shape();
+        for (size_t i = 0; i < output_placements.size(); ++i) {
+            if (i < mesh_shape.dims() && mesh_shape[static_cast<int>(i)] > 1) {
+                output_placements[i] = shard_placement;
+            } else {
+                output_placements[i] = tt::tt_metal::distributed::MeshMapperConfig::Replicate{};
+            }
         }
     }
 

